### PR TITLE
fix #2497 Time did not support microseconds and nanoseconds properly,…

### LINKED ISSUE #2497: 5.0: Time needs to properly support different time units
There needs to be some solid unit tests for time parsing and it should handle all these cases (same ones supported by elastic server side:

https://github.com/exceptionless/Exceptionless.DateTimeExtensions/blob/master/tests/Exceptionless.DateTimeExtensions.Tests/TimeUnitTests.cs

I bring this up because I did `new Time("-3h").ToTimeSpan()` and it returned a timespan of `-3ms` which is completely wrong, it should be `-3 hours`.

### PR DESCRIPTION
**Pull request by Mpdreamz**
… our parsing also failed in many cases where es wouldn't, added support for MINUS_ONE and ZERO.

### CODE REVIEW

File: src/Nest/CommonOptions/TimeUnit/Time.cs [hunk 1]
```diff
@@ -16,57 +16,99 @@ public class Time : IComparable<Time>, IEquatable<Time>
...
 
-		public Time(TimeSpan timeSpan)
-		{
-			Reduce(timeSpan.TotalMilliseconds);
-		}
+		public static Time MinusOne { get; } = new Time(-1, true);
+		public static Time Zero { get; } = new Time(0, true);
 
-		public Time(double milliseconds)
+		protected Time(int specialFactor, bool specialValue)
```
Comment (russcam) [inline src/Nest/CommonOptions/TimeUnit/Time.cs]: Make `private`

File: src/Nest/CommonOptions/TimeUnit/Time.cs [hunk 2]
```diff
@@ -16,57 +16,99 @@ public class Time : IComparable<Time>, IEquatable<Time>
 		private const double MillisecondsInAnHour = MillisecondsInAMinute * 60;
 		private const double MillisecondsInAMinute = MillisecondsInASecond * 60;
 		private const double MillisecondsInASecond = 1000;
+		private const double NanosecondsInAMillisecond = 100;
+		private const double MicrosecondsInAMillisecond = 10;
```
Comment (russcam) [inline src/Nest/CommonOptions/TimeUnit/Time.cs]: 1000 microseconds in a millisecond. I think this should be `TicksPerMicrosecond`

File: src/Nest/CommonOptions/TimeUnit/Time.cs [hunk 3]
```diff
@@ -16,57 +16,99 @@ public class Time : IComparable<Time>, IEquatable<Time>
 		private const double MillisecondsInAnHour = MillisecondsInAMinute * 60;
 		private const double MillisecondsInAMinute = MillisecondsInASecond * 60;
 		private const double MillisecondsInASecond = 1000;
+		private const double NanosecondsInAMillisecond = 100;
```
Comment (russcam) [inline src/Nest/CommonOptions/TimeUnit/Time.cs]: 1,000,000 nanoseconds in a millsecond. I think this should be `NanosecondsInATick`

File: src/Nest/CommonOptions/TimeUnit/Time.cs [hunk 4]
```diff
@@ -16,57 +16,99 @@ public class Time : IComparable<Time>, IEquatable<Time>
...
+		private const double MicrosecondsInAMillisecond = 10;
 
-		private static readonly Regex _expressionRegex = new Regex(@"^(?<factor>[-+]?\d+(?:\.\d+)?)(?<interval>(?:y|M|w|d|h|m|s|ms))?$", RegexOptions.Compiled | RegexOptions.ExplicitCapture);
+		private static readonly Regex ExpressionRegex =
+			new Regex(@"^(?<factor>[-+]?\d+(?:\.\d+)?)\s*(?<interval>(?:y|w|d|h|m|s|ms|nanos|micros))?$",
+				RegexOptions.Compiled | RegexOptions.ExplicitCapture | RegexOptions.IgnoreCase);
```
Comment (russcam) [inline src/Nest/CommonOptions/TimeUnit/Time.cs]: I would keep this a case-sensitive match and include `M` for Months, along with the parsing `ToEnum<TimeUnit>()` as case-sensitive; this better reflects what's accepted by Elasticsearch

File: src/Nest/CommonOptions/TimeUnit/Time.cs [hunk 5]
```diff
@@ -16,57 +16,99 @@ public class Time : IComparable<Time>, IEquatable<Time>
...
-					? match.Groups["interval"].Value.ToEnum<TimeUnit>(StringComparison.Ordinal)
-					: TimeUnit.Millisecond;
+				case null:
+					throw new ArgumentException($"Time expression '{timeUnit}' is missing an interval", nameof(timeUnit));
+				case "M":
+					this.Interval = TimeUnit.Month;
+					break;
+				case "m":
+					this.Interval = TimeUnit.Minute;
+					break;
```
Comment (russcam) [inline src/Nest/CommonOptions/TimeUnit/Time.cs]: See [above comment](#r99513122)

File: src/Tests/CommonOptions/TimeUnit/TimeUnits.doc.cs [hunk 6]
```diff
@@ -100,14 +102,76 @@ public void EqualityAndComparable()
 			(twoDays <= new Time("2d")).Should().BeTrue();
 
 			/**
+			* Special Time values `0` and `-1` can be compared against eachother
+			* and other Time values although admittingly this is a tad nonsensical.
```
Comment (russcam) [inline src/Tests/CommonOptions/TimeUnit/TimeUnits.doc.cs]: admittedly

File: src/Nest/CommonOptions/TimeUnit/Time.cs (hunk 4 above)
Comment (Mpdreamz) [inline src/Nest/CommonOptions/TimeUnit/Time.cs]: Elasticsearch does case insensitive, e.g NANOS, MICROS are all accepted.

**Comment by Mpdreamz** [comment]
@forloop updated the PR with feedback, I didn't drop the internals down to nanoseconds because we expose milliseconds and I fear the implementation becoming even more confusing. 

Mind taking a crack at it?

File: src/Nest/CommonOptions/TimeUnit/Time.cs [hunk 7]
```diff
@@ -132,23 +197,31 @@ public bool Equals(Time other)
...
 			if (ReferenceEquals(this, obj)) return true;
 			if (obj.GetType() != this.GetType()) return false;
-			return Equals((Time)obj);
+			return Equals((Time) obj);
 		}
 
-		public override int GetHashCode() => this.ApproximateMilliseconds.GetHashCode();
+		public override int GetHashCode()
+		{
+			if (this.StaticTimeValue.HasValue) return this.StaticTimeValue.Value;
```
Comment (gmarz) [inline src/Nest/CommonOptions/TimeUnit/Time.cs]: Yes, I think so

**Comment by Mpdreamz** [comment]
merged to `5.x`

**Comment by jonyadamit** [comment]
Hey guys, changing enums this way, requires at least some notification, preferably all capitals, in a **Breaking Changes** section for the release.  
I was quite took by surprise to see that query information saved in our database suddenly changed it's meaning (a monthly date histogram was changed to daily for instance...).
In any way, keep up the good work 😄

**Pull request merged**